Fixes issue with MultiTenant,fails to retrieve parent image from…

egorbatik opened the pull request:
first time it tries to bring it from the parent site.

This is from Module Version 2.1.x as is the first compatible with
Broadleaf 5.2


**A Brief Overview**
Add image in global, change to site, delete temporary directory, avoid browser cache (like refreshing or going back and forward from global to site again), global image will not be retrieved.

It should bring from S3, currently it was assuming the image in S3 was in the same site, this fix will ensure if it fails the first time it will try to look for the parent

**Link to QA issue**
https://github.com/BroadleafCommerce/QA/issues/3643

**How to Test**
1. Get Private Demo In Local Environment with a DB.
2. Add the Amazon Module in pom.xml (In core)
```
   <!-- BroadLeaf Amazon -->
        <dependency>
    		<groupId>org.broadleafcommerce</groupId>
    		<artifactId>broadleaf-amazon</artifactId>
		</dependency>
````
3. Set the configuration in the following way (in common-shared, the S3 Bucket is mine):
````
asset.server.url.prefix.internal=cmsstatic
asset.server.url.prefix.secure=https://s3.us-east-2.amazonaws.com/broadleaftest/img/


aws.s3.accessKeyId=AKIAJEAAMYC5ZDPPMJGA
aws.s3.secretKey=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

aws.s3.defaultBucketName=broadleaftest
aws.s3.bucketSubDirectory=img

aws.s3.defaultBucketRegion=us-east-2
````
4. Now to test the current bug:
4.1 Add Image from Global (from Media)
4.2 Change to Marketplace (or any site)
4.3 Add a image. (From Media menu)
4.4 Delete from /tmp (java.tmp.io) directory every image.
4.5 Change to Master Site
4.6 Change to Marketplace
4.7. Go to Media Menu and you will the global image is missing. (Some times the image is cached in the browser, but this usually fix that, if not try again)

Is very important when going back to Master Site TO NOT GO to media menu because it will write again the images in the tmp directory invalidating the Test.


5. To test the fix, load the modified version of the module and try the same steps as 4.

egorbatik committed:
MultiTenant Support for Amazon Module, when fails to retrieve image the
first time it tries to bring it from the parent site.

This is from Module Version 2.1.x as is the first compatible with
Broadleaf 5.2

bpolster commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -77,37 +83,72 @@
+     * @param name
+     * @throws RuntimeException
+     */
+    private void ensureFileCreation(File returnFile,String name) throws RuntimeException
```
Use standard formatting the opening { should be on the same line as the filename.   Ensure all new code uses this convention.

egorbatik replied:
Resolved in Next Commit

bpolster commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +386,113 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
     }
+    
+    @Override
+	public File getResource(String name, FileApplicationType fileApplicationType) {
```
Use standard spacing - indent to 4 spaces, use 4 spaces instead of tabs.   This should be a setting in your IDE

egorbatik replied:
Done in IDE, spaces resolved in next commit

bpolster commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +386,113 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+ * Gets getParentId method by reflection   
+ * @return
+ */
+	private Method getMultiTenantGetParentSiteIdMethod() {
```
BLC framework convention is to use protected instead of private.   Allows users of the framework to more easily self-serve.

egorbatik replied:
Resolved in next commit

bpolster commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +386,113 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+					ReflectionUtils.invokeMethod(getMultiTenantGetParentSiteIdMethod(), site).toString());
+			return result;
+		} catch ( SecurityException | IllegalArgumentException e) {
+			// TODO Auto-generated catch block
```
Log this exception.   Generally, never bury an exception.

egorbatik replied:
Logged (Logging logic added), resolved in next commit. Always return valid states.

bpolster commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -130,13 +171,48 @@ public File getResource(String name, FileApplicationType fileApplicationType) {
-        }
-        return defaultBucketName;
+    
+    public File getResourceDefault(String rawName, FileApplicationType fileApplicationType) { 
```
getResourceDefault and getResourceFromParent share a lot of the logic.     Refactor so that they can share this common code.

egorbatik replied:
One single method for both. only adding one parameter. Resolved in Next commit

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -77,37 +83,72 @@
+    
+    private static final String MULTITENANT_SITE_CLASSNAME= "com.broadleafcommerce.tenant.domain.MultiTenantSite";
+    
+    private static final String MULTITENANTSITE_GETPARENTID_METHODNAME= "getParentSiteId";
```
I don't think you should keep this as a constant

egorbatik replied:
From now one will protected a NOT final. Next Commit

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +386,113 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+			if (site != null) {
+				//The MultiTenantSite.getParentSiteId method is called by reflection.
+				String siteDirectory = getSiteDirectory((invokeMultiTenantGetParentSiteIdMethod(site)));
+				resourceName= StringUtils.removeStart(resourceName, "/");
```
I think that you don't need this because you're using `FilenameUtils.concat`, right?

egorbatik replied:
It also came from inherited code, but removed. Resolved in next commit

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +386,113 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+    @Override
+	public File getResource(String name, FileApplicationType fileApplicationType) {
+		File file = getResourceDefault(name, fileApplicationType);
+		if (file.getPath().contains("this/path/should/not/exist/")  && isMultiTenantEnvironment()) {
```
Don't use this hardcoded string. Instead check `file.exists()`

egorbatik replied:
This came from the original code, the original method was returning this in this way, I know does not look good but I'm not sure if it is is being used outside this module.
I think cannot use file.exists() because this file may no exists in the file system.

phillipuniverse reviewed (changes requested):
Just a few additional comments, but generally formatting like @bpolster mentioned.

Also, this module already has some good automated integration tests that ensure that things are retrieved from S3 ok. It would be better to try to make use of that here as well. To test the differences between MT/non-MT you could subclass this in your test and override the `isMultiTenantEnvironment()` method to always return true or false depending on which way you're testing. I don't think I would _require_ these additional tests (since I'm not sure what the support priority is for getting in these changes) to merge this in, but it is worth some effort.

egorbatik committed:
Changes requested by PR.

Spaces
Opening brackets same line.
Refactor to do shared logic in one single method.
Exception handling, Logging but always returning valid states

egorbatik committed:
Code review, changes applied.
* Logs (Log Level Debug)
* Refactors

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +371,124 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+            }
+            if (site != null) {
+                //The MultiTenantSite.getParentSiteId method is called by reflection.
+                String siteDirectory = getSiteDirectory((invokeMultiTenantGetParentSiteIdMethod(site)));
```
Don't you need to gate this call? If for some reason this invocation failed, you are hardcoding in a result of `0L`, which means you'll look in a directory for `site-0`. Can't you just skip this altogether if there was a problem getting to the parent site id?

egorbatik replied:
Agree. Fixed.

egorbatik committed:
Gate removed.

egorbatik committed:
Skipping altogether, if there is a problem with the class, will return
the resource as it comes, if there is a problem with the Parent (As
somehow cannot be retrieved from the DB, will return 0).

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +371,121 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+                //this exception should never occur if the MultiTenant environment is ok
+               LOG.error("Problem reflecting Multitenant Class", e);
+            }
+            //The MultiTenantSite.getParentSiteId method is called by reflection.
```
Well now this comment is out of date and you should remove it

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +371,121 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+            Object site="";
+            try {
+                site = (getMultiTenantClass().cast(brc.getNonPersistentSite()));
+                String siteDirectory = getSiteDirectory((invokeMultiTenantGetParentSiteIdMethod(site)));
```
You still have the exact same problem here. If there is no parent, you are going to read zero, which means you will look for a `site-0` directory.

egorbatik replied:
If site-0 (and parent) will bring info from S3

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +371,121 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+    protected Long invokeMultiTenantGetParentSiteIdMethod(Object site) {
+        try {
+            Long result =Long.parseLong(
+                    ReflectionUtils.invokeMethod(getMultiTenantGetParentSiteIdMethod(), site).toString());
```
This can return null if there is no parent site id, which I believe will throw an NPE in this case.

egorbatik replied:
Retrieving and checking for null in next commit.

egorbatik committed:
If is looking for parent but the parent has a not a valid ID it will not
go to S3.

egorbatik committed:
Avoiding NPE in case there is no parentId

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -77,39 +85,102 @@
-            S3Object object = s3.getObject(new GetObjectRequest(bucketName, buildResourceName(name)));
+            String bucketName = getBucketName(rawName, s3config.getDefaultBucketName());
+            name = getResourceName(s3config,bucketName, rawName);
+            boolean skip=false;
```
What is this doing here?

egorbatik replied:
Fixed

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -306,4 +374,123 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+            } catch (ClassNotFoundException | LinkageError e) {
+                //The MultitenantEnvironment should be checked before reaching this point so
+                //this exception should never occur if the MultiTenant environment is ok
+               LOG.error("Problem reflecting Multitenant Class", e);
```
You should not be erroring here. It is a perfectly valid case to not find the multitenant class to get the parent id, if multitenant is not on your classpath.

egorbatik replied:
Yes, as I'm asking in previous step if this class exists with isMultiTenantEnvironment I should never get trapped into this exception... but....
1. In a previous comment, Brian said we never bury any exception.

2. Let's suppose someone is extending this class and is somehow bypassing the isMultiTenantEnvironment call. In that case it makes sense to log the exception.

phillipuniverse replied:
Ok, I guess I see that now but it was pretty confusing for me to trace down.

As long as you have absolutely tested these changes in the DemoSite project (where no multitenancy exists at all) and you do not get any extraneous error logs, then we are good to go.

egorbatik replied:
In Demo site is working, I don't see errors. Is worth to do a double check.
I used DemoSite solution as base, as it has no multitenant (and is reflected in S3 as well).
I don't know if Demo Site has a limitation I should consider.

phillipuniverse replied:
Cool, as long as you have validated that everything works fine in demo site with the S3 upload (sounds like you validated it on the S3 side) and you can load the image in a browser, then we are good to merge.

charchar replied:
@egorbatik Can you confirm that this is good to merge based on @phillipuniverse's  comment?

egorbatik replied:
Good to merge

phillipuniverse reviewed (changes requested):
Ok last couple of things and we can call this good enough. Did you validate this running Broadleaf _without_ multi-tenant? like if you hook this into https://github.com/BroadleafCommerce/DemoSite (no enterprise, no multi-tenant at all) does it still work ok?

egorbatik committed:
skip variable removed as is not needed.

phillipuniverse commented:
@egorbatik you can consider this approved pending your verification with the DemoSite project.
